pkg/util/kubernetes/kubelet/kubelet.go

```diff
@@ -402,72 +409,76 @@ func (ku *KubeUtil) GetRawMetrics() ([]byte, error) {
 
-	// HTTP
-	ku.kubeletApiEndpoint = fmt.Sprintf("http://%s:%d", ku.kubeletHost, config.Datadog.GetInt("kubernetes_http_kubelet_port"))
-	_, code, httpUrlErr := ku.QueryKubelet(kubeletPodPath)
-	if httpUrlErr == nil {
-		if code == http.StatusOK {
-			log.Debugf("Kubelet endpoint is: %s", ku.kubeletApiEndpoint)
-			return nil
+	for i, h := range potentialKubeletHosts {
+		log.Infof("Host %d/%d, trying to use host %s with HTTPS and HTTP settings", i+1, len(potentialKubeletHosts), h)
+		ku.kubeletHost = h
+		err := ku.setupKubeletApiClient()
+		if err != nil {
+			return err
 		}
-		return fmt.Errorf("unexpected status code %d on endpoint %s%s", code, ku.kubeletApiEndpoint, kubeletPodPath)
-	}
-	log.Debugf("Cannot query %s%s: %s", ku.kubeletApiEndpoint, kubeletPodPath, httpUrlErr)
 
-	return fmt.Errorf("cannot connect: https: %q, http: %q", httpsUrlErr, httpUrlErr)
-}
+		// HTTPS first
+		ku.kubeletApiEndpoint = fmt.Sprintf("https://%s:%d", ku.kubeletHost, config.Datadog.GetInt("kubernetes_https_kubelet_port"))
+		log.Debugf("Trying to query the kubelet endpoint %s ...", ku.kubeletApiEndpoint)
+		response, err := c.Get(ku.kubeletApiEndpoint + "/")
+		if err == nil {
+			log.Infof("Successfully queried to %s without any security settings, adding security transport settings to query %s%s", response.Request.URL, ku.kubeletApiEndpoint, kubeletPodPath)
 
-func (ku *KubeUtil) init() error {
-	var err error
+			response, err := ku.doKubeletRequest(kubeletPodPath)
+			if err == nil {
+				log.Infof("Successfully connected securely on kubelet endpoint %s", response.Request.URL)
+				switch {
+				case response.StatusCode == http.StatusOK:
+					log.Infof("Successfully authorized to query the kubelet on %s: 200, using %s as kubelet endpoint", response.Request.URL, ku.kubeletApiEndpoint)
+					return nil
 
-	// setting the kubeletHost
-	ku.kubeletHost = config.Datadog.GetString("kubernetes_kubelet_host")
-	if ku.kubeletHost == "" {
-		ku.kubeletHost, err = docker.HostnameProvider("")
-		if err != nil {
-			return fmt.Errorf("unable to get hostname from docker, please set the kubernetes_kubelet_host option: %s", err)
-		}
-	}
+				case response.StatusCode >= http.StatusInternalServerError:
+					log.Infof("Unexpected return code on request %s on kubelet endpoint %s", response.Request.URL, ku.kubeletApiEndpoint)
 
-	// Trying connectivity insecurely with a dedicated client
-	c := http.Client{Timeout: time.Second}
-	c.Transport = &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
+				case response.StatusCode == http.StatusUnauthorized:
+					log.Debugf("Unauthorized to request %s on kubelet endpoint %s, check the kubelet authentication/authorization settings")
+
+				default:
+					log.Debugf("Unexpected http code %d on kubelet endpoint %s", response.StatusCode)
+				}
+
+				// err != nil
+			} else if strings.Contains(err.Error(), "x509: certificate is valid for") {
```

Comment:
This makes sense, but I feel like the logs are the first place where customers and support engineers are looking to to troubleshoot these issues, and not directly the `agent diagnose`, which they are usually less familiar with. That can allow quicker resolutions, or even help the customer to avoid reaching the support there.
However, this should not stop us, as this can be handled with doc or KBs.